### DISCUSSION TITLE
writer validation not validating against Implementation xml?

### DISCUSSION THREAD

**Discussion by d-floyd** [xlate/staedi#284] [Q&A]
I'm running validation with the xml stream writer -- but it seem only be validating using the transaction.xml and not the implementation.xml
For example:
I have this in my Transaction.xml
```
                     .....
            <loop code="Loop1000" minOccurs="1" maxOccurs="4" >
                <sequence>
                    <segment type="N1"/>
                    <loop code="Loop1100" maxOccurs="1">
                        <sequence >
                            <segment type="ACT" minOccurs="1" />
                        </sequence>
                    </loop>
                </sequence>
            </loop>
            <loop name="Loop2000"  code="Loop2000" type="Loop2000"  maxOccurs="9999">
                <sequence>
                    <segment type="INS" minOccirs="1"/>
                     .....

```
and i have this, in  my Implementaion.xml (I have a B and C Loop1000 as well.. with a minOccurs=0, the A loop  is the only one required)
```
            <loop  title="Loop1000A" typeId="Loop1000A" code="Loop1000A" type="Loop1000" discriminator="1" minOccurs="1" maxOccurs="1">
                <sequence>
                    <segment type="N1" title="Sponsor Name"  minOccurs="1" description="1000">
                        <sequence>
                            <element position="1" title="Entity Identifier Code (Sponsor)" minOccurs="1">
                                <enumeration>
                                    <value>P5</value>
                                </enumeration>
                            </element>
                            <element position="2" minOccurs="0"/>
                            <element position="3" minOccurs="1"/>
                            <element position="4" minOccurs="1"/>
                        </sequence>
                    </segment>
                </sequence>
            </loop>
```

when I validate my transaction that has the B and C loop -- validation passes ,unexpectedly, since there is no matching 1000A loop.
if I remove the B and C loops as well.. then it fails. Not exactly as expected.. since   I get the error on the next segment
Encountered SEGMENT_ERROR [MANDATORY_SEGMENT_MISSING] in segment INS  --> the INS segment exists, its the first segment after the 1000 loop... it's the Loop1000 that's missing... but at least it fails .

My question is -- should the validation be performed using the implementation schema? Or is my expectation not correct?
Should the validation message indicate that it's the Loop1000 that's missing (as opposed to the INS)?

**Comment by MikeEdgar** [comment]
@d-floyd - just to confirm, when you create the schema, you load it via `Implementation.xml` correct? If so, I would expect it to validate the A, B, and C loops separately. Otherwise, this will require some additional digging.

**Reply by d-floyd** [reply]
I do yes..
where schema is String variable containing "schema/834/Implementation.xml" 

            if (!StringUtils.isAllBlank(schema)) {
                ediWriter.setControlSchema(SchemaFactory.newFactory()
                    .getControlSchema("X12", new String[]{"00501"}));
                ediWriter.setTransactionSchema(getSchema(schema)); 

            }

    public Schema getSchema(String schemaName) {
        ClassLoader classLoader = getClass().getClassLoader();
        SchemaFactory schemaFactory;
        schemaFactory = SchemaFactory.newFactory();
        InputStream inputStream = classLoader.getResourceAsStream(schemaName);
        try {
            return schemaFactory.createSchema(inputStream);
        } catch (EDISchemaException ex) {
            String msg = String.format("Cannot read schema: %s", schemaName);
            log.error(msg);
            throw new UserMessageException(msg, ex);
        }

    }

I verified that this works fine for the reader.. but not the writer.

**Reply by d-floyd** [reply]
here is a simplified test...  hopefully it will help in your diagnosis (or to confirm that I've done something wrong)
The test method  reads in a "good" file -- passes as expected
reads in a file missing the 1000X loop --> errors as expected.
reads the same bad file w/o the validatoin -- to the generated xml
attempts to write that "bad" xml --> success -- expected this to fail since it's missing the 1000A loop. verified the edi stream isn't valid, it doesn't have the 1000A loop. The only difference between the bad generated edi string and the "good" edi string from the file. Is  the missing 1000A loop.
[ValidationTest.zip](https://github.com/xlate/staedi/files/9135207/ValidationTest.zip)

**Reply by MikeEdgar** [reply]
Thank you for the reproducer - I'll take a look hopefully later today or tomorrow morning.

**Reply by MikeEdgar** [reply]
Pardon the delay. I'm still planning to go through this.

**Reply by MikeEdgar** [reply]
This is going to require an enhancement to work. The writer is not dealing with the implementation schema the same as the reader.

**Reply by d-floyd** [reply]
OK. Thanks for looking into it (and confirming it wasn't something on my end). It's easy enough to work around with a java validation on object we create. Thanks for looking into it.
